Src/Base/AMReX_Array4.H

```diff
@@ -56,212 +57,373 @@ namespace amrex {
+            constexpr_for<0, dim>([&] (auto d) {
+                s *= (end[d] - begin[d]);
+            });
+            return s;
         }
 
+        template <typename... idx, std::enable_if_t<detail::ArrayNDIndexCheck_impl_v<dim, idx...>,int> = 0>
         [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-        std::size_t size () const noexcept {
-            return this->nstride * this->ncomp;
+        bool contains (idx... i) const noexcept {
+            if constexpr (sizeof...(i) == dim) {
+                return this->contains(IntVectND<dim>{i...});
+            } else {
+                // When number of indices is one less than dimension, we assume the last index is zero
+                return this->contains(IntVectND<dim>{i...,0});
+            }
         }
 
         [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-        int nComp () const noexcept { return ncomp; }
+        bool contains (IntVect const& iv) const noexcept {
+#if (AMREX_SPACEDIM == 1)
+            return this->contains(IntVectND<dim>{iv[0],0,0,0});
+#elif (AMREX_SPACEDIM == 2)
+            return this->contains(IntVectND<dim>{iv[0],iv[1],0,0});
+#elif (AMREX_SPACEDIM == 3)
+            return this->contains(IntVectND<dim>{iv[0],iv[1],iv[2],0});
+#endif
+        }
 
         [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-        bool contains (int i, int j, int k) const noexcept {
-            return (i>=begin.x && i<end.x && j>=begin.y && j<end.y && k>=begin.z && k<end.z);
+        bool contains (IntVectND<dim> const& iv) const noexcept {
+            bool inside = true;
+            constexpr_for<0, dim>([&] (auto d) {
+                inside = inside && (iv[d] >= begin[d]) && (iv[d] < end[d]);
+            });
+            return inside;
         }
 
+        template <int Dim=dim, std::enable_if_t<Dim==4,int> = 0>
         [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-        bool contains (IntVect const& iv) const noexcept {
-            return AMREX_D_TERM(   iv[0]>=begin.x && iv[0]<end.x,
-                                && iv[1]>=begin.y && iv[1]<end.y,
-                                && iv[2]>=begin.z && iv[2]<end.z);
+        bool contains (Dim3 const& cell) const noexcept {
+            return this->contains(IntVectND<dim>{cell.x, cell.y, cell.z, 0});
         }
 
+        template <int Dim=dim, std::enable_if_t<Dim==4,int> = 0>
         [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-        bool contains (Dim3 const& cell) const noexcept {
-            return this->contains(cell.x,cell.y,cell.z);
+        CellData<T> cellData (int i, int j, int k) const noexcept {
+            int ncomp = end[3] - begin[3];
+            return CellData<T>(this->ptr(i,j,k), nstride[dim-2], ncomp);
+        }
+
+
+        template <typename... idx, std::enable_if_t<sizeof...(idx) == dim,int> = 0>
+        [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+        constexpr Long get_offset (idx... indices) const noexcept
```

Comment:
I think we can simplify this and remove this version of `get_offset`by constructing an IntVectND at the place where it is called. Then `get_impl` can also be removed.